var/spack/repos/builtin.mock/packages/nosource/package.py

```diff
@@ -0,0 +1,17 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+from spack import *
+
+
+class Nosource(BundlePackage):
+    """Simple bundle package with one optional dependency"""
+
+    homepage = "http://www.example.com"
+
+    version('1.0', '0123456789abcdef0123456789abcdef')
```

Comment:
> This makes me think of another question -- what happens if a package doesn't have a `version()` directive at all? 

At this point the ``NoValidVersionError`` is raised when an attempt is made to concretize the Spec.

Reply:
I removed the hash and the tests seem to pass.  I'm currently waiting for ``xsdk`` to build without its version hash.

Reply:
According to the code comments, the purpose for the ``version`` directive (or at least its metadata) is for fetching the source code.  This would seem to imply ``version`` should not even be allowed.  (Or the comments should be changed to reflect another purpose, such as documenting the version history of the bundle.)

Reply:
> This shouldn't require a checksum -- it doesn't make sense if there's nothing to download. We should make the checksum arg optional and validate it in `PackageMeta` after directives are processed. That part of the code is kind of nasty so lemme know if you want to look at it together.

Thanks for the offer.  I believe I figured out the earliest place in `PackageMeta` that I could enforce the checksum constraint in my commit yesterday.

> 
> This makes me think of another question -- what happens if a package doesn't have a `version()` directive at all? I am not sure we actually handle that case anywhere. Seems like we should ensure there is at least one version (so that the concretizer can always pick something). That, or raise an error if there are no `version()` directives and no version is passed on the command line. The latter is more flexible but the check would need to be done pre-concretization, instead of in `PackageMeta`. I _think_ I like the latter better (as it supports all the ways Spack allows you to specify a version), but I have not thought it through.

I experimented a little yesterday with a `version()` directive check and the introduction of a mock-builtin package with no directive.  Adding the mock package to the tests -- albeit with a check in `PackageMeta` -- resulted in numerous failed tests involving `Spec.concretize` for different specs.  The problem occurred when loading the mock package module as part of `concretize`'s `_expand_virtual_packages`.

I'm not sure I understand your "...need to be done pre-concretization" point so could use more information on the process.